### PR TITLE
Introduce unit tests in GitHub CI workflow

### PR DESCRIPTION
**Pull request by sayerhs**

### CODE REVIEW

File: .github/workflows/linux.yml [hunk 1]
```diff
@@ -49,6 +55,9 @@ jobs:
         python3 -m pip install -U pip
```
Comment (ax3l) [inline .github/workflows/linux.yml]:
```suggestion
        python3 -m pip install -U pip pytest
```

File: .github/workflows/linux.yml [hunk 2]
```diff
@@ -14,9 +14,12 @@ jobs:
       run: .github/workflows/dependencies/dependencies.sh
     - name: Build & Install
       run: |
-        python3 -m pip install -U pip setuptools wheel
+        python3 -m pip install -U pip setuptools wheel pytest
         AMREX_MPI=ON python3 -m pip install -v .
         python3 -c "import pyamrex; print(pyamrex.__version__)"
+    - name: Unit tests
+      run: |
+        pytest tests/
```
Comment (ax3l) [inline .github/workflows/linux.yml]: Not like this for maximum portability?
```suggestion
        mpirun -n 2 python3 -m pytest tests/
```

File: .github/workflows/linux.yml [hunk 3]
```diff
@@ -30,9 +33,12 @@ jobs:
       run: |
         export CC=$(which gcc-10)
         export CXX=$(which g++-10)
-        python3 -m pip install -U pip setuptools wheel
+        python3 -m pip install -U pip setuptools wheel pytest
         AMREX_MPI=ON python3 -m pip install -v .
         python3 -c "import pyamrex; print(pyamrex.__version__)"
+    - name: Unit tests
+      run: |
+        pytest tests/
```
Comment (ax3l) [inline .github/workflows/linux.yml]:
```suggestion
        mpirun -n 2 python3 -m pytest tests/
```

File: .github/workflows/linux.yml [hunk 4]
```diff
@@ -49,6 +55,9 @@ jobs:
         python3 -m pip install -U pip
         python3 -m pip install -v .
         python3 -c "import pyamrex; print(pyamrex.__version__)"
+    - name: Unit tests
+      run: |
+        pytest tests/
```
Comment (ax3l) [inline .github/workflows/linux.yml]:
```suggestion
        python3 -m pytest tests/
```

File: .github/workflows/macos.yml [hunk 5]
```diff
@@ -15,8 +15,12 @@ jobs:
       run: .github/workflows/dependencies/dependencies_mac.sh
     - name: Build & Install
       run: |
-        python3 -m pip install -U pip
+        python3 -m pip install -U pip pytest
         python3 -m pip install -v .
         python3 -c "import pyamrex; print(pyamrex.__version__)"
+    - name: Unit tests
+      run: |
+        pytest tests/
```
Comment (ax3l) [inline .github/workflows/macos.yml]:
```suggestion
        python3 -m pytest tests/
```

File: .github/workflows/windows.yml [hunk 6]
```diff
@@ -13,9 +13,12 @@ jobs:
         python-version: '3.x'
     - name: Build & Install
       run: |
-        python -m pip install -U pip
+        python -m pip install -U pip pytest
         python -m pip install -v .
         python -c "import pyamrex; print(pyamrex.__version__)"
+    - name: Unit tests
+      run: |
+        pytest tests/
```
Comment (ax3l) [inline .github/workflows/windows.yml]:
```suggestion
        python -m pytest tests/
```

File: README.md [hunk 7]
```diff
@@ -120,6 +127,15 @@ From the base of the pyAMReX source directory, execute:
 python3 -m pip install -v .
 ```
 
+On successful installation, you can run the unit tests (assuming `pytest` is
+installed)
+
+```bash
+# Run all tests 
+pytest tests/
```
Comment (ax3l) [inline README.md]:
```suggestion
python3 -m pytest tests/
```

File: README.md [hunk 8]
```diff
@@ -120,6 +127,15 @@ From the base of the pyAMReX source directory, execute:
...
+```
+
```
Comment (ax3l) [inline README.md:139]:
```suggestion
```

File: README.md [hunk 9]
```diff
@@ -62,6 +62,7 @@ Optional dependencies include:
 - [mpi4py](https://www.openmp.org) 2.1+: for multi-node and/or multi-GPU execution
 - [CCache](https://ccache.dev): to speed up rebuilds (needs 3.7.9+ for CUDA)
 - further [optional dependencies of AMReX](https://github.com/AMReX-Codes/amrex/)
+- [pytest](https://docs.pytest.org/en/stable/): for running unit tests
```
Comment (ax3l) [inline README.md]: Let's add a reasonable lower bar like 6+ or 6.2+ or so:
```suggestion
- [pytest](https://docs.pytest.org/en/stable/) 6.2+: for running unit tests
```

File: README.md [hunk 10]
```diff
@@ -120,6 +127,15 @@ From the base of the pyAMReX source directory, execute:
 python3 -m pip install -v .
 ```
 
+On successful installation, you can run the unit tests (assuming `pytest` is
+installed)
```
Comment (ax3l) [inline README.md]: For small diffs, I usually write one sentence per line in `.md` and `.rst` (like in LaTeX):
```suggestion
On successful installation, you can run the unit tests (assuming `pytest` is installed)
```

File: .github/workflows/linux.yml [hunk 11]
```diff
@@ -46,9 +52,12 @@ jobs:
       run: |
         export CC=$(which clang)
         export CXX=$(which clang++)
-        python3 -m pip install -U pip
+        python3 -m pip install -U pip pytest
         python3 -m pip install -v .
         python3 -c "import pyamrex; print(pyamrex.__version__)"
+    - name: Unit tests
+      run: |
+        mpiexec -np 1 python3 -m pytest tests/
```
Comment (ax3l) [inline .github/workflows/linux.yml]: This unit test has no MPI (MPI is off by default for Python atm.):
```suggestion
        python3 -m pytest tests/
```

File: .github/workflows/linux.yml [hunk 12]
```diff
@@ -14,9 +14,12 @@ jobs:
       run: .github/workflows/dependencies/dependencies.sh
     - name: Build & Install
       run: |
-        python3 -m pip install -U pip setuptools wheel
+        python3 -m pip install -U pip setuptools wheel pytest
         AMREX_MPI=ON python3 -m pip install -v .
         python3 -c "import pyamrex; print(pyamrex.__version__)"
+    - name: Unit tests
+      run: |
+        python3 -m pytest tests/
```
Comment (ax3l) [inline .github/workflows/linux.yml]:
```suggestion
        mpiexec -np 1 python3 -m pytest tests/
```

Comment (ax3l) [inline .github/workflows/linux.yml]: Until we add `mpi4py`:
```suggestion
#    - name: Unit tests
#      run: |
#        mpiexec -n 1 python3 -m pytest tests/
```

File: .github/workflows/linux.yml [hunk 13]
```diff
@@ -30,9 +33,12 @@ jobs:
       run: |
         export CC=$(which gcc-10)
         export CXX=$(which g++-10)
-        python3 -m pip install -U pip setuptools wheel
+        python3 -m pip install -U pip setuptools wheel pytest
         AMREX_MPI=ON python3 -m pip install -v .
         python3 -c "import pyamrex; print(pyamrex.__version__)"
+    - name: Unit tests
+      run: |
+        mpiexec -np 1 python3 -m pytest tests/
```
Comment (ax3l) [inline .github/workflows/linux.yml:41]:
```suggestion
#    - name: Unit tests
#      run: |
#        mpiexec -np 1 python3 -m pytest tests/
```

File: .github/workflows/windows.yml [hunk 14]
```diff
@@ -13,9 +13,12 @@ jobs:
         python-version: '3.x'
     - name: Build & Install
       run: |
-        python -m pip install -U pip
+        python -m pip install -U pip pytest
         python -m pip install -v .
         python -c "import pyamrex; print(pyamrex.__version__)"
+    - name: Unit tests
+      run: |
+        python -m pytest tests/
```
Comment (ax3l) [inline .github/workflows/windows.yml:21]: Windows paths are `\`, so let's just skip it?
```suggestion
        python -m pytest tests
```

File: src/Base/AMReX.cpp [hunk 15]
```diff
@@ -45,4 +45,13 @@ void init_AMReX(py::module& m)
     m.def("finalize", py::overload_cast<AMReX*>(&Finalize));
 
     m.def("space_dim", [](){ return AMREX_SPACEDIM; });
+
+    m.def("have_mpi",
```
Comment (ax3l) [inline src/Base/AMReX.cpp:49]: actually that would make a great property
```suggestion
    m.def_property_readonly("have_mpi",
```

File: tests/conftest.py [hunk 16]
```diff
@@ -3,6 +3,9 @@
 import pytest
 import pyamrex as amrex
 
+if amrex.have_mpi():
```
Comment (ax3l) [inline tests/conftest.py:6]:
```suggestion
if amrex.have_mpi:
```

File: src/Base/AMReX.cpp (hunk 15 above)
Comment (sayerhs) [reply src/Base/AMReX.cpp:49]: @ax3l I don't think you can do `def_property_readonly` on a module object.

Comment (ax3l) [reply src/Base/AMReX.cpp:49]: true, sorry!

**Review by ax3l — APPROVED** [review]
:rocket:

**Pull request merged**